### ISSUE TITLE
C# - Spikes in Native Memory Usage

### ISSUE THREAD

**Issue by smflorentino** [grpc/grpc#20352]
### What version of gRPC and what language are you using?
C#, 1.20.1

### What operating system (Linux, Windows,...) and version?
Windows Server 2016

### What runtime / compiler are you using (e.g. python version or version of gcc)
.NET Framework 4.5

### What did you do?
After a process restart, our service suddenly experiencing DeadlineExceeded errors, and more importantly, tons of Native memory usage. I'm not sure which is causing the other, but the native memory spikes do happen at the same time as the DeadlineExceeded errors.

Usage eventually goes down, but spikes of this magnitude cause us to run of of memory. Can this be tuned? Is there memory being held in native that should be released sooner? Or maybe I need to adjust timeout settings?
### What did you expect to see?
I didn't expect to see such spikes of native memory use.

### What did you see instead?

There are 2 errors in the logs that I see. The first one that I know generally happens if the server tries to send a response when the client is no longer listening.

```
[gRPC Warning] Exception while handling RPC. System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at Grpc.Core.Internal.AsyncCallServer`*.SendStatusFromServerAsync(Status status, Metadata trailers, Nullable`* optionalWrite)
   at Grpc.Core.Internal.ClientStreamingServerCallHandler`*.<HandleCall>d__*.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Grpc.Core.Server.<HandleCallAsync>d__*.MoveNext()
```

The second error I am not familiar with, but it's not actionable. How can more useful information be surfaced?
```
[gRPC Warning] Exception while handling RPC. System.IO.IOException: Error sending status from server.
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Grpc.Core.Internal.ClientStreamingServerCallHandler`*.<HandleCall>d__*.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Grpc.Core.Server.<HandleCallAsync>d__*.MoveNext()
```

Here are our metric showing gRPC server error codes, CPU, and Native/Managed memory use:
![grpc](https://user-images.githubusercontent.com/3495150/65545476-4e4dea80-deca-11e9-944c-3eecf92cf778.PNG)

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.

**Comment by jtattermusch** [comment]
@smflorentino  would you be able to provide code that reproduces the problem?
Does this happen regularly or on some rare occasions?

Would it be possible to experiment with newer version of gRPC? (we've made some changes regarding memory management in the last few releases).

You're right that one of the log messages is harmless https://github.com/grpc/grpc/issues/9377
about the other once I'm not sure either.

**Comment by jtattermusch** [comment]
Can you also describe the situation in which you get the deadline exceeded error? Is the memory usage coming from a client or from a server (do you run both client and server in the same process?)

**Comment by smflorentino** [comment]
I finally had some time to get back to root causing this.

>would you be able to provide code that reproduces the problem?
>Does this happen regularly or on some rare occasions?

This happens rarely, and in production only, but I think I found the trigger.

As soon as I call `await this.server.ShutdownAsync()`, native memory uses grows for several minutes until the system runs out of memory and the process crashes. I will say, we have clients still attempting to call the server *after* `ShutdownAsync` was called (those calls time out with `DeadlineExceeded`). Could that be the cause of this?

**Comment by jtattermusch** [comment]
@smflorentino thanks for the inputs. Since you know the characteristics of the behavior and you have some pointers when the memory spikes happen, would you be able to come up with a minimal repro
(a server on which you call server.ShutdownAsync() at some point and clients trying to talk to it).
These conditions are hard for me to simulate as I don't know the pattern in which you invoke RPCs (what type, what message size, what frequency, ..) , how often you shutdown servers shutdown, how many client/servers you have. If you can come up with a reproduction, I can definitely try to see what the root cause is.

Talking in general -  invoking server.ShutdownAsync() should definitely not be a reason for memory consumption to increase indefinitely.
The behavior should be:
1. the existing RPCs continue to be served normally. 
2. newly attempted RPC are rejected (the DeadlineExceeded timeout seems odd - I'll try to reproduce that at least).

**Comment by jtattermusch** [comment]
In my quick experiment, after calling `var shutdownTask = server.ShutdownAsync();`, all the new RPC I start fail immediately with `StatusCode.Unavailable` (= new call are rejected by the server), is that not what you're seeing?

**Comment by smflorentino** [comment]
Hey @jtattermusch thanks very much for your response. We're using bidirectional streaming RPC. I can also try to do a console repro with CloseAsync.

We just actually had this happen again last night, but I realized there's another native component, and it could be that one, too. I need write some infrastructure to capture memory allocation ETW events automatically in Production nodes that see this issue, but that will take me some time. Once I have those stacks, I should be able to determine the call stack of the allocation, and therefore the DLL.

I hope to work on this in the next few weeks. And we should probably update gRPC, as well.

**Comment by markdroth** [comment]
Closing this pending more information from the user.

**Comment by smflorentino** [comment]
@jtattermusch I finally got some time to get back to this. I updated to the latest verison, v2.29.1, and I still see this happening. I added some more instrumentation, and I found that `IAsyncStreamReader<T>.MoveNext()` is taking an incredibly long time to complete; over 4 minutes in some cases. This is a S2S RPC call - and the client nodes are healthy. Because `MoveNext` takes so long, there is a high # of *concurrent* calls and thus the spike in native memory usage (I am assuming it's due to the native library buffering those calls).

I still can't find a repro, since a node in an otherwise healthy state suddenly starts having issues without warning.

I do find it suspect that `IAsyncStreamReader<T>` takes 4 minutes to complete. Especially because:
- The deadline is 1 minute
- I log the `ServerCallConext.CancellationToken.IsCancellationRequested` *after* `MoveNext` completed, and the token is indeed cancelled, but what is strange is that the `MoveNext` call still completed successfully (and likely ignored the token).

I suspect 2 issues:
- There is a `CancellationToken` being ignored
- Some kind of lock contention / thread starvation, but it's not clear what triggers this. The node (4 core, 14GB RAM) had < 1000 rps before this started happening.

Does this help narrow the problem?
